# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [5]:
import sqlite3
from pickle import HIGHEST_PROTOCOL

import pandas as pd
import warnings

from pygments import highlight
from win32netcon import ACCESS_GROUP

warnings.simplefilter('ignore')

Zet het bestand go_sales_train.sqlite in een makkelijk te vinden map

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales_train.sqlite te maken.

In [6]:
conn = sqlite3.connect('../Makkelijk_te_vinden_map/go_sales_train.sqlite')

Lees van de ingelezen go_sales_train-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [7]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
tables = pd.read_sql_query(sql_query, conn)
tables

type           name       tbl_name  rootpage  \
0   table        country        country       480   
1   table  order_details  order_details        11   
2   table   order_header   order_header        24   
3   table   order_method   order_method        27   
4   table        product        product        28   
5   table   product_line   product_line        17   
6   table   product_type   product_type        10   
7   table  retailer_site  retailer_site        18   
8   table  return_reason  return_reason        30   
9   table  returned_item  returned_item        31   
10  table   sales_branch   sales_branch         2   
11  table    sales_staff    sales_staff         3   

                                                  sql  
0   CREATE TABLE "country" (\n\t"COUNTRY_CODE"\tIN...  
1   CREATE TABLE "order_details" (\n\t"ORDER_DETAI...  
2   CREATE TABLE "order_header" (\n\t"ORDER_NUMBER...  
3   CREATE TABLE "order_method" (\n\t"ORDER_METHOD...  
4   CREATE TABLE "product" (\n\t"PRODUCT_NUMBER"\t...  
5   CREATE TABLE "product_line" (\n\t"PRODUCT_LINE...  
6   CREATE TABLE "product_type" (\n\t"PRODUCT_TYPE...  
7   CREATE TABLE "retailer_site" (\n\t"RETAILER_SI...  
8   CREATE TABLE "return_reason" (\n\t"RETURN_REAS...  
9   CREATE TABLE "returned_item" (\n\t"RETURN_CODE...  
10  CREATE TABLE "sales_branch" (\n\t"SALES_BRANCH...  
11  CREATE TABLE "sales_staff" (\n\t"SALES_STAFF_C...

Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt. <u>Let op:</u> lees eventueel de informatie uit het Notebook van werkcollege 1.1b nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [8]:
sql_query = "SELECT name FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
Naam_tabellen = pd.read_sql(sql_query, conn)
#Op de puntjes hoort de connectie naar go_sales_train óf go_staff_train óf go_crm_train te staan.
Naam_tabellen

name
0         country
1   order_details
2    order_header
3    order_method
4         product
5    product_line
6    product_type
7   retailer_site
8   return_reason
9   returned_item
10   sales_branch
11    sales_staff

erachter 

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [9]:
sql_query = "SELECT * FROM product ;"
products = pd.read_sql_query(sql_query, conn)
products_dataframe = pd.DataFrame(products)

products_tussen_100_en_50 = (products_dataframe['PRODUCTION_COST'] >=50) & (products_dataframe['PRODUCTION_COST'] <=100)

products_dataframe.loc[products_tussen_100_en_50, ['PRODUCT_NAME','PRODUCTION_COST']]

PRODUCT_NAME  PRODUCTION_COST
5          TrailChef Deluxe Cook Set            85.11
7             TrailChef Double Flame            75.00
16                   Hibernator Lite            60.00
17                        Hibernator            86.00
19   Hibernator Self - Inflating Mat            78.55
22               Hibernator Camp Cot            65.33
23      Canyon Mule Climber Backpack            62.50
45           Granite Climbing Helmet            52.86
47             Husky Harness Extreme            55.00
57                    Granite Hammer            57.13
68                         Polar Ice            73.33
69                      Polar Sports            80.00
70                        Polar Wave            60.00
71                     Polar Extreme            72.50
74                      Edge Extreme            80.00
77                         Seeker 35            79.19
78                         Seeker 50            92.58
79                    Seeker Extreme            94.12
82                    Glacier Deluxe            56.25
83                       Glacier GPS            78.55
109                Blue Steel Putter            60.00
110            Blue Steel Max Putter            81.80
113              Course Pro Golf Bag            80.00

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [10]:
products_margin_ding = (products_dataframe['MARGIN'] < 0.2) | (products_dataframe['MARGIN'] >0.6)

products_dataframe.loc[products_margin_ding, ['PRODUCT_NAME', 'MARGIN']]

PRODUCT_NAME  MARGIN
20                Hibernator Pad    0.17
23  Canyon Mule Climber Backpack    0.17
85             BugShield Natural    0.67
86               BugShield Spray    0.67
87         BugShield Lotion Lite    0.70
88              BugShield Lotion    0.63
89             BugShield Extreme    0.63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [11]:
sql_query = "SELECT * FROM country ;"
country = pd.read_sql_query(sql_query, conn)
country_dataframe = pd.DataFrame(country)

country_francs = country_dataframe['CURRENCY_NAME'] == 'francs'

country_dataframe.loc[country_francs, ['COUNTRY']]

COUNTRY
0        France
7   Switzerland
15      Belgium

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [12]:
introductiedatums_meer_dan_50 = products_dataframe['MARGIN'] > 0.5

introductie = products_dataframe.loc[introductiedatums_meer_dan_50, ['INTRODUCTION_DATE']]
unieke_introductie = introductie.drop_duplicates()
unieke_introductie

INTRODUCTION_DATE
66         2000-10-26
85         1995-02-15
101        2003-12-10
103        2003-12-18
104        2003-12-27
106        2004-01-13
110        2003-12-15

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [13]:
sql_query = "SELECT * FROM sales_branch ;"
sales_branch = pd.read_sql_query(sql_query, conn)
sales_branch_dataframe = pd.DataFrame(sales_branch)
verkoopafdeling = (sales_branch_dataframe['REGION'].notna()) & (sales_branch_dataframe['ADDRESS2'].notna())
sales_branch_dataframe.loc[verkoopafdeling, ['ADDRESS1', 'CITY']]


ADDRESS1              CITY
2                Singelgravenplein 4         Amsterdam
13  Prol. Paseo de la Reforma No. 51  Distrito Federal
14               202-2-3 Hyakunincho             Tokyo
15                     543-225 Asahi        Osaka City
16                  2315 Queen's Ave         Melbourne
18             Avenida Paulista, 333         São Paulo
24                    3 Albany Court        Birmingham

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [14]:
landen_met_dollars = (country_dataframe['CURRENCY_NAME'] == 'dollars') | (country_dataframe['CURRENCY_NAME'] == 'new dollar')
country_dataframe.loc[landen_met_dollars, ['COUNTRY']]

COUNTRY
2   United States
3          Canada
11         Taiwan
14      Australia

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [15]:
sql_query = "SELECT * FROM retailer_site ;"
retailer_site = pd.read_sql_query(sql_query, conn)
retailer_site_dataframe = pd.DataFrame(retailer_site)
postcode_met_D = (retailer_site_dataframe['POSTAL_ZONE'].str[0] == 'D' ) & (retailer_site_dataframe['ADDRESS2'].notna() )
retailer_site_dataframe.loc[postcode_met_D, ['ADDRESS1', 'ADDRESS2', 'CITY']]

ADDRESS1       ADDRESS2       CITY
31  Röntgenstraße 90  3. Tür rechts  Frankfurt
59   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [16]:
sql_query = "SELECT * FROM returned_item;"
returned_item = pd.read_sql_query(sql_query, conn)
returned_item_dataframe = pd.DataFrame(returned_item)
returned_item_dataframe['RETURN_QUANTITY'].sum()

np.int64(11908)

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [17]:
sales_branch_dataframe['REGION'].notna().count()

np.int64(28)

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [18]:
laagste_marge_products = products_dataframe['MARGIN'].min()
hoogste_marge_products = products_dataframe['MARGIN'].max()
gemiddelde_marge_products = products_dataframe['MARGIN'].mean()

resultaat = pd.DataFrame({
    "laasgte marge products" : [round(laagste_marge_products, 2)],
    "hoogste marge products" : [round(hoogste_marge_products, 2)],
    "gemiddelde marge products" : [round(gemiddelde_marge_products,2)]
})

resultaat

laasgte marge products  hoogste marge products  gemiddelde marge products
0                    0.17                     0.7                        0.4

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [19]:
geen_2e_address = retailer_site_dataframe['ADDRESS2'].isna()
resultaat = retailer_site_dataframe.loc[geen_2e_address, 'CITY'].count()
resultaat

np.int64(286)

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [20]:
sql_query = "SELECT * FROM order_details ;"
order_details = pd.read_sql_query(sql_query, conn)
order_details_dataframe = pd.DataFrame(order_details)

producten_met_korting =    order_details_dataframe['UNIT_SALE_PRICE'] < order_details_dataframe['UNIT_PRICE']
order_details_dataframe.loc[producten_met_korting,'UNIT_SALE_PRICE'].mean().round(2)

np.float64(180.44)

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [21]:
sql_query = "SELECT * FROM sales_staff ;"
sales_staff = pd.read_sql_query(sql_query, conn)
sales_staff_dataframe = pd.DataFrame(sales_staff)

opdr = sales_staff_dataframe.groupby('POSITION_EN', as_index=False)['SALES_STAFF_CODE'].count()

opdr.rename(columns={'SALES_STAFF_CODE':'aantal medewerkers'}, inplace=True)
opdr

POSITION_EN  aantal medewerkers
0                Branch Manager                  19
1              District Manager                   4
2               General Manager                   1
3  Level 1 Sales Representative                  13
4  Level 2 Sales Representative                  39
5  Level 3 Sales Representative                  24
6              Regional Manager                   2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [22]:
opdr = sales_staff_dataframe.groupby('WORK_PHONE', as_index=False)['SALES_STAFF_CODE'].count()
opdr_filter = opdr[opdr['SALES_STAFF_CODE']> 4]
opdr_filter.rename(columns={'SALES_STAFF_CODE':'aantal medewerkers bereikbaar'}, inplace=True)
opdr_filter

WORK_PHONE  aantal medewerkers bereikbaar
4     +(44) 121 3505267                              5
10  +(61) 03 2982  4242                              5
11  +31 (0)20 692 93 94                              5
13    +33 1 68 94 52 20                              5
18     1 (206) 292-0012                              5
19     1 (305) 557-4810                              6
20     1 (310) 281-5722                              5
21     1 (403) 232-5986                              5
22     1 (416) 493-5595                              5
23     1 (617) 268-6754                              6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [23]:
land_merge_retailer = pd.merge(retailer_site_dataframe,country_dataframe, left_on= 'COUNTRY_CODE', right_on= 'COUNTRY_CODE', how = 'outer')

land_merge_retailer.loc[land_merge_retailer['COUNTRY'] == 'Netherlands', ['ADDRESS1', 'CITY']]

ADDRESS1          CITY
215                   Beets 36         Beets
216               Perestraat 1      Deventer
217  Laan van Meerdervoort 966      Den Haag
218             Kalfjeslaan 47    Amstelveen
219               Startbaan 11    Amstelveen
220      Noorderspoorsingel 28     Groningen
221              Hogehilweg 13     Amsterdam
222             Spoorstraat 43    Varsseveld
223             Westhavenweg 8     Amsterdam
224              Westzeedijk 4        Muiden
225            Gooiergracht 91         Laren
226            Hoofdstraat 150   Pieterburen
227                Olde Diek 2         Laren
228     Amsterdamsestraatweg 8      Maarssen
229        Krayenhoffstraat 11     Amsterdam
230         Overschiestraat 19     Amsterdam
231              Havenbuurt 12        Marken
232               Duivendaal 1    Wageningen
233              Voorstraat 37  Nieuweschans
234         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [30]:
sql_query = "SELECT * FROM product_type ; "
product_type = pd.read_sql_query(sql_query, conn)
product_type_dataframe = pd.DataFrame(product_type)
product_merge_type= pd.merge (products_dataframe, product_type_dataframe, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how = 'outer')
product_merge_type.loc[product_merge_type['PRODUCT_TYPE_EN'] == 'Eyewear', 'PRODUCT_NAME']

67        Polar Sun
68        Polar Ice
69     Polar Sports
70       Polar Wave
71    Polar Extreme
Name: PRODUCT_NAME, dtype: object

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [71]:
sql_query = "SELECT * FROM order_header ;"
order_header = pd.read_sql_query(sql_query, conn)
order_header_dataframe = pd.DataFrame(order_header)

order_merge_retailer = pd.merge(order_header_dataframe, retailer_site_dataframe, left_on='RETAILER_SITE_CODE', right_on='RETAILER_SITE_CODE', how = 'inner')

order_retailer_merge_staff = pd.merge(order_merge_retailer, sales_staff, left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how = 'inner')

order_retailer_merge_staff.loc[order_retailer_merge_staff['POSITION_EN'] == 'Branch Manager', ['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']].drop_duplicates()

ADDRESS1 FIRST_NAME LAST_NAME
3185  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 

In [84]:
staff_merge_order = pd.merge(sales_staff_dataframe, order_header_dataframe, left_on= 'SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how = 'outer')

staff_merge_order.loc[staff_merge_order['POSITION_EN'].str.contains('Manager'), ['POSITION_EN', 'ORDER_DATE']].drop_duplicates()

POSITION_EN  ORDER_DATE
0       Branch Manager         NaN
320   Regional Manager         NaN
648   District Manager         NaN
3748    Branch Manager  2024-09-09
3749    Branch Manager  2024-03-13
3750    Branch Manager  2024-01-09
4186   General Manager         NaN

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [94]:
sql_query = "SELECT * FROM order_details ;"
order_details = pd.read_sql_query(sql_query, conn)
order_details_dataframe = pd.DataFrame(order_details)

product_merge_order_details= pd.merge(order_details_dataframe, products_dataframe, left_on= 'PRODUCT_NUMBER', right_on= 'PRODUCT_NUMBER', how = 'inner')

product_order_details_type = pd.merge(product_merge_order_details, product_type_dataframe, left_on='PRODUCT_TYPE_CODE', right_on='PRODUCT_TYPE_CODE', how = 'inner')

product_order_details_type.loc[product_order_details_type['QUANTITY'] > 750, ['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates()


PRODUCT_NAME    PRODUCT_TYPE_EN
539        BugShield Extreme  Insect Repellents
1759                Star Peg              Tents
20121  BugShield Lotion Lite  Insect Repellents
20147         Sun Shelter 15          Sunscreen
20152         Sun Shelter 30          Sunscreen
31236        BugShield Spray  Insect Repellents
37272              Firefly 2           Lanterns
37348       BugShield Lotion  Insect Repellents
37363             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [97]:
korting = (product_merge_order_details['UNIT_PRICE'] - product_order_details_type['UNIT_SALE_PRICE'])/product_order_details_type['UNIT_PRICE']

product_merge_order_details.loc[korting > 0.4, ['PRODUCT_NAME']].drop_duplicates()

PRODUCT_NAME
506              BugShield Natural
511                BugShield Spray
650   Hailstorm Titanium Woods Set
2797         BugShield Lotion Lite
2836                Sun Shelter 30
3751             BugShield Extreme
5074                Sun Shelter 15
6930              BugShield Lotion

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [109]:
sql_query = "SELECT * FROM returned_item;"
returned_item = pd.read_sql_query(sql_query, conn)
returned_item_dataframe = pd.DataFrame(returned_item)

returned_merge_details = pd.merge(returned_item_dataframe, order_details_dataframe, left_on='ORDER_DETAIL_CODE', right_on='ORDER_DETAIL_CODE', how = 'inner')

retour= returned_merge_details['RETURN_QUANTITY']/returned_merge_details['QUANTITY']

sql_query = "SELECT * FROM return_reason;"
return_reason = pd.read_sql_query(sql_query, conn)
return_reason_dataframe = pd.DataFrame(return_reason)

return_reason_details = pd.merge(returned_merge_details, return_reason_dataframe, left_on='RETURN_REASON_CODE', right_on='RETURN_REASON_CODE', how = 'inner')

return_reason_details.loc[retour > 0.9, 'RETURN_DESCRIPTION_EN'].drop_duplicates()

0    Unsatisfactory product
2     Wrong product shipped
3     Wrong product ordered
Name: RETURN_DESCRIPTION_EN, dtype: object

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [112]:
product_merge_type.groupby('PRODUCT_TYPE_EN', as_index=False)['PRODUCT_TYPE_CODE'].count()

PRODUCT_TYPE_EN  PRODUCT_TYPE_CODE
0             Binoculars                  4
1   Climbing Accessories                  7
2           Cooking Gear                 10
3                Eyewear                  5
4              First Aid                  5
5       Golf Accessories                  4
6      Insect Repellents                  5
7                  Irons                  4
8                 Knives                  5
9               Lanterns                 12
10            Navigation                  4
11                 Packs                  6
12               Putters                  3
13                  Rope                  4
14                Safety                  4
15         Sleeping Bags                  7
16             Sunscreen                  5
17                 Tents                  6
18                 Tools                  6
19               Watches                  5
20                 Woods                  4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [118]:
retailer_country= pd.merge(retailer_site_dataframe, country_dataframe, left_on='COUNTRY_CODE', right_on='COUNTRY_CODE', how='outer')

retailer_country.groupby('COUNTRY', as_index=False)['COUNTRY_CODE'].size()


COUNTRY  size
0        Australia    10
1          Austria    10
2          Belgium    10
3           Brazil     1
4           Canada    31
5            China    10
6          Denmark     6
7          Finland     5
8           France    29
9          Germany    32
10           Italy    12
11           Japan    25
12           Korea    10
13          Mexico    10
14     Netherlands    20
15           Spain     9
16          Sweden    12
17     Switzerland    13
18          Taiwan    10
19  United Kingdom    25
20   United States   101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen)

In [138]:
cooking_gear_orders = product_order_details_type.loc[product_order_details_type['PRODUCT_TYPE_EN']=='Cooking Gear']

antw =cooking_gear_orders.groupby(['PRODUCT_NAME'], as_index=False).agg(totaal_verkocht=('QUANTITY', 'sum'), gemiddelde_prijs=('UNIT_SALE_PRICE', 'mean'))

antw.sort_values(by='totaal_verkocht', ascending=False)


PRODUCT_NAME  totaal_verkocht  gemiddelde_prijs
9        TrailChef Water Bag            36738          6.032506
2              TrailChef Cup            27418          6.976490
0          TrailChef Canteen            26544         11.979418
6      TrailChef Kitchen Kit            22206         22.345975
8         TrailChef Utensils            13776         17.294151
4     TrailChef Double Flame            12476        132.672882
1         TrailChef Cook Set            10878         51.687926
5           TrailChef Kettle             8826         12.201235
7     TrailChef Single Flame             8504         63.827622
3  TrailChef Deluxe Cook Set             3110        121.217638

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [140]:
# Merge verkoopafdelingen met landnamen
sales_branch_merge = pd.merge(sales_branch_dataframe, country_dataframe, on='COUNTRY_CODE', how='left')

# Selecteer de stad van de verkoopafdeling en hernoem naar 'verkoper'
sales_branch_merge = sales_branch_merge[['COUNTRY', 'CITY']].rename(columns={'CITY': 'verkoper'})

# Merge klantvestigingen met landnamen
retailer_site_merge = pd.merge(retailer_site_dataframe, country_dataframe, on='COUNTRY_CODE', how='left')

# Tel het aantal unieke steden per land waar klanten zich bevinden
klanten_per_land = retailer_site_merge.groupby('COUNTRY', as_index=False)['CITY'].nunique()
klanten_per_land.rename(columns={'CITY': 'klanten'}, inplace=True)

# Merge de twee datasets om per land de verkoper en het aantal klantsteden te krijgen
resultaat = pd.merge(sales_branch_merge, klanten_per_land, on='COUNTRY', how='left')

resultaat

COUNTRY          verkoper  klanten
0           France             Paris       14
1            Italy            Milano        9
2      Netherlands         Amsterdam       15
3          Germany           Hamburg       18
4          Germany           München       18
5           Sweden             Kista       10
6           Canada           Calgary       16
7           Canada           Toronto       16
8    United States            Boston       62
9    United States           Seattle       62
10   United States       Los Angeles       62
11   United States             Miami       62
12          France              Lyon       14
13          Mexico  Distrito Federal        6
14           Japan             Tokyo        5
15           Japan        Osaka City        5
16       Australia         Melbourne        6
17           Spain            Bilbao        5
18          Brazil         São Paulo        1
19         Finland            Kuopio        5
20           Korea             Seoul        6
21          Taiwan         San Chung        9
22           China          Shanghai        7
23  United Kingdom            London        9
24  United Kingdom        Birmingham        9
25     Switzerland            Zürich        9
26         Belgium          Heverlee        8
27         Austria              Wien        9

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [144]:
# Merge sales_staff_dataframe met order_details_dataframe om te kijken wie iets heeft verkocht
sales_with_orders = pd.merge(sales_staff_dataframe, order_header_dataframe, left_on='SALES_STAFF_CODE', right_on='SALES_STAFF_CODE', how='left')

# Filter medewerkers die nooit iets hebben verkocht
never_sold = sales_with_orders[sales_with_orders['_merge'] == 'left_only'][['FIRST_NAME', 'LAST_NAME']].drop_duplicates()

# Resultaat weergeven
never_sold

FIRST_NAME      LAST_NAME
0         Denis           Pagé
320        Else         Mörike
321       Frank          Fuchs
647       Fritz         Hirsch
648        Jörg          Kunze
728       Maria      Iacobucci
946        Kick        Kalkman
1309      Karin      Bergström
1395      Sally          White
1421      Frank        Bretton
1422    Georges  Saint-Germain
1719       Alex      Rodriguez
2088   Samantha          Floyd
2392      Julie          Olsen
2530      Corey         Wright
2873    Estelle        Johnson
2874       Greg        Belding
2999      Julie        Béchard
3157      Lucía          Reyna
3401     Kazumi        Uragome
3559      Alice         Walter
4185     Walter         Taylor
4186   Penelope        Tamrine
4349      Polly         Linton
4590      Gracy        Gellens

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [153]:
# Bereken de gemiddelde marge van alle producten
gemiddelde_marge = products_dataframe['MARGIN'].mean()

# Filter producten waarvan de marge lager is dan de gemiddelde marge
producten_lagere_marge = products_dataframe[products_dataframe['MARGIN'] < gemiddelde_marge]

# Bereken het aantal producten en de gemiddelde marge van deze groep
aantal_producten = len(producten_lagere_marge)
gemiddelde_marge_lager = producten_lagere_marge['MARGIN'].mean()

# Maak een dataframe met het resultaat
resultaat = pd.DataFrame({
    'Waarden': [aantal_producten, round(gemiddelde_marge_lager, 2)]
}, index=['Aantal producten', 'Gemiddelde marge'])


resultaat


Waarden
Aantal producten    59.00
Gemiddelde marge     0.29

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [154]:
# Filter producten met een verkoopprijs hoger dan 500
producten_onder500 = order_details_dataframe[order_details_dataframe['UNIT_SALE_PRICE'] > 500]

all_returned = returned_item_dataframe['ORDER_DETAIL_CODE'].dropna().unique()
# Filter producten die nooit zijn teruggebracht (RETURN_QUANTITY is 0 of NaN)
producten_nooit_geretourneerd = producten_onder500.loc[~producten_onder500['ORDER_DETAIL_CODE'].isin(all_returned), ['PRODUCT_NUMBER']]

products_dataframe.loc[products_dataframe['PRODUCT_NUMBER'].isin(producten_nooit_geretourneerd.squeeze()), ['PRODUCT_NAME']]


PRODUCT_NAME
11                           Star Dome
12                        Star Gazer 2
13                        Star Gazer 3
14                        Star Gazer 6
44                      Husky Rope 200
100              Hailstorm Steel Irons
101           Hailstorm Titanium Irons
102         Lady Hailstorm Steel Irons
103      Lady Hailstorm Titanium Irons
104       Hailstorm Titanium Woods Set
105          Hailstorm Steel Woods Set
106  Lady Hailstorm Titanium Woods Set
107     Lady Hailstorm Steel Woods Set

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [155]:
medewerkers= sales_staff_dataframe.loc[:,['LAST_NAME','POSITION_EN']]

for index, medewerker in medewerkers.iterrows():
    rol = medewerkers.at[index, 'POSITION_EN']
    if "Manager" in rol:
        medewerkers.at[index, 'manager'] = "Ja"
    else:
        medewerkers.at[index, 'manager'] = "Nee"

medewerkers= medewerkers.drop('POSITION_EN', axis=1)
medewerkers

LAST_NAME manager
0         Pagé      Ja
1       Michel     Nee
2     Clermont     Nee
3       Jauvin     Nee
4    Wiesinger     Nee
..         ...     ...
97    Laermans     Nee
98     De Crée     Nee
99     Lattrez     Nee
100  Seefelder     Nee
101     Grüner     Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [156]:
from datetime import date
date.today().year

2025

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [157]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [159]:
medewerkers = sales_staff_dataframe.loc[:,['LAST_NAME','DATE_HIRED']]

huidig_jaar = date.today().year

for index, medewerker in medewerkers.iterrows():
    date_hired = medewerkers.at[index, 'DATE_HIRED']
    date_str = date_hired
    date_format = '%Y-%m-%d'
    date_obj = datetime.strptime(date_str, date_format)

    aantal_jaar_in_dienst = huidig_jaar - date_obj.year

    if aantal_jaar_in_dienst < 25:
        medewerkers.at[index, 'tijd in dienst'] = "kort in dienst"
    else:
        medewerkers.at[index, 'tijd in dienst'] = "lang in dienst"

medewerkers = medewerkers.drop('DATE_HIRED', axis=1)
medewerkers


LAST_NAME  tijd in dienst
0         Pagé  lang in dienst
1       Michel  lang in dienst
2     Clermont  lang in dienst
3       Jauvin  lang in dienst
4    Wiesinger  lang in dienst
..         ...             ...
97    Laermans  lang in dienst
98     De Crée  lang in dienst
99     Lattrez  lang in dienst
100  Seefelder  lang in dienst
101     Grüner  lang in dienst

[102 rows x 2 columns]

## Van Jupyter Notebook naar Pythonproject

1. Richt de map waarin jullie tot nu toe hebben gewerkt in volgens de mappenstructuur uit de slides.
2. Maak van de ontstane mappenstructuur een Pythonproject dat uitvoerbaar is vanuit de terminal. Maak daarin een .py-bestand dat minstens 5 antwoorden uit dit notebook (in de vorm van een DataFrame) exporteert naar Excelbestanden. Alle notebooks mogen als notebook blijven bestaan.
3. Zorg ervoor dat dit Pythonproject zijn eigen repo heeft op Github. Let op: je virtual environment moet <b><u>niet</u></b> meegaan naar Github.

Je mag tijdens dit proces je uit stap 1 ontstane mappenstructuur aanpassen, zolang je bij het beoordelingsmoment kan verantwoorden wat de motivatie hierachter is. De slides verplichten je dus nergens toe.